### PR TITLE
Session interaction refactoring

### PR DESCRIPTION
**Pull request by mbabker**
#### Summary of Changes

To try and sum things up:
- `JSession` can now lazy start a session; calls to `get()`, `set()`, `has()`, and `clear()` will start the session if it hasn't explicitly been started previously
- The application classes are restructured to not explicitly start the session; this defers the point `JSession::start()` is called from the application constructors to one of the first actions done after `$app->execute()` is called.  The advantage here is that now sessions are started after the application has been instantiated and set to `JFactory::$application`.
- The logic for purging of session metadata is altered so that this step only occurs if the database session handler is not in use and it is now executed as part of the `onAfterRender` event instead of when the session is originally started; garbage collection will occur for the database session handler as part of the PHP session API
- `JApplicationCms::checkSession()` now optionally accepts the `JSession` object that should be checked and this will be a mandatory argument as of 4.0.  This method is called as part of the `afterSessionStart` event which may occur without a session object stored to `JFactory`.
- The `afterSessionStart` event now includes the `JSession` object for which the session was started for as an argument of the event, this should be used instead of fetching the session from `JFactory`
- Adjustments to the database session handler:
  - `JSessionStorageDatabase` is dependent on the application to insert the session record into the database table, the logic is added to the `write()` method to remove this dependency
  - Garbage collection when the database session store is used is deferred to the end of the request cycle to help prevent a request failure solely because expired session records are being purged.  As explained in the [PHP documentation](http://php.net/manual/en/sessionhandler.gc.php) the `gc()` method is called when the session is started.
  - `JApplicationCms::checkSession()` is adjusted to only manipulate the `time` column when not using the database session handler to preserve the record's integrity when written by the session API
#### Testing Instructions

1) Download the branch from https://github.com/mbabker/joomla-cms/archive/refactor-session-metadata-coupling.zip
2) Install Joomla; sessions in the installation application should continue to function correctly
3) Use Joomla normally; sessions should continue to function correctly
4) With the database session handler enabled, `JApplicationCms::checkSession()` should only be executed when the session is new.  As this is only ensuring the extra metadata is present in the database, it is not required for every page load.  All requests after the session is started will have the record in the database updated by the session handler only (or any place making an explicit query to the `#__session` database table).  As well, the new `JApplicationCms::cleanupSessionMetadata()` method should never be called.
5) With any other session handler enabled, `JApplicationCms::checkSession()` should only be called if the session is new or the time is an even numbered second.  The new `JApplicationCms::cleanupSessionMetadata()` method should be called when the `onAfterRender` event is triggered and the delete query should only run when the time is an even numbered second.

### CODE REVIEW

**Comment by pritalpatel** [comment]
I have followed the steps provided. I successfully tested 1-3 steps. But not able to understand 4th and 5th step clearly. I am sure this is me who don’t understand as I don’t much familiar with joomla framework so far.

Thank you. :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10905">issues.joomla.org/joomla-cms/10905</a>.</sub>

**Comment by mbabker** [comment]
Look, I know this is tricky as all can be to properly test, but there are major logic and workflow fixes introduced here that are somewhat needed to break some existing bad practices.  Some tests here would be appreciated...

File: libraries/cms/application/cms.php [hunk 1]
```diff
@@ -156,65 +169,148 @@ public function afterSessionStart()
...
+	 */
+	public function cleanupSessionMetadata()
+	{
+		$time = time();
+
+		if ($time % 2)
+		{
+			$db = JFactory::getDbo();
+
+			// The modulus introduces a little entropy, making the flushing less accurate but fires the query less than half the time.
```
Comment (PhilETaylor) [inline libraries/cms/application/cms.php]: move this up to line 289 ;-)

File: installation/application/web.php [hunk 2]
```diff
@@ -50,6 +50,23 @@ public function __construct()
 	}
 
 	/**
+	 * Event listener for the `onAfterSessionStart` event.
+	 *
+	 * @param   JSession  $session  Session object
+	 *
+	 * @return  void
+	 *
+	 * @since   3.6
```
Comment (andrepereiradasilva) [inline installation/application/web.php]: __DEPLOY_VERSION__ exists now ;)

Comment (PhilETaylor) [inline installation/application/web.php]: > **DEPLOY_VERSION** exists now ;)

And is probably documented nowhere apart from inline in build.php... :)

Comment (mbabker) [inline installation/application/web.php]: Joomla have good documentation?  What dream world are you in?  😜

Comment (PhilETaylor) [inline installation/application/web.php]: Ive been in symfony land for a month... loving having excellent documentation :)

File: tests/unit/suites/libraries/cms/html/JHtmlMenuTest.php [hunk 3]
```diff
@@ -17,6 +17,38 @@
 class JHtmlMenuTest extends TestCaseDatabase
 {
 	/**
+	 * Sets up the fixture, for example, opens a network connection.
+	 * This method is called before a test is executed.
+	 *
+	 * @return  void
+	 *
+	 * @since   3.6
```
Comment (andrepereiradasilva) [inline tests/unit/suites/libraries/cms/html/JHtmlMenuTest.php]: you missed this two ;)

File: libraries/joomla/application/web.php [hunk 4]
```diff
@@ -1081,8 +1081,26 @@ public function loadSession(JSession $session = null)
 	 *
 	 * @since   12.2
 	 */
-	public function afterSessionStart(JSession $session)
+	public function afterSessionStart(JSession $session = null)
```
Comment (PhilETaylor) [inline libraries/joomla/application/web.php]: #TIL that PHP sucks and than you can type hint the class, and set a null, and PHP will allow @session to be null|JSession ... which in mind mind sucks, as null ==! JSession :-) :-)

Comment (mbabker) [inline libraries/joomla/application/web.php]: I take it you haven't seen the nullable types RFC that was passed for PHP 7.1?  https://wiki.php.net/rfc/nullable_types

Comment (PhilETaylor) [inline libraries/joomla/application/web.php]: How to find time to live a life and still find time to learn all this stuff, new stuff and old stuff.... honestly - I work over 12 hours most days, 7 days a week most weeks and still dont find the time to keep up with code changes....

**Comment by mbabker** [comment]
Is this going anywhere or can I close for lack of interest?

**Comment by andrepereiradasilva** [comment]
from my part i find all your PR interresting but because of gsoc project lack of time ATM.

**Comment by PhilETaylor** [comment]
Im back at my desk on monday and will play with this for you - although with my history of testing things ... :)

**Comment by mbabker** [comment]
Out of sync yet again and apparently too complex for the user base.  I'm done.

**Comment by PhilETaylor** [comment]
This has only been open 4 months, even some simpler PRs have been open since Aug 18, 2014 ;-)

**Comment by mbabker** [comment]
After the discussion yesterday I'm keeping my PRs limited to things that a user can click in the backend after installing a patch with patch tester since apparently requesting testing on highly technical matters such as this is asking for too much.

**Comment by andrepereiradasilva** [comment]
please don't close your PR michael, i think most of your PR are important.

i don't have always time to test all of them, but i try to test them when possible.

**Comment by mbabker** [comment]
I will try this one last time.  Key points here:

- The database handler has a hardcoded dependency to the record being inserted by the application, this is now fixed
- The database metadata cleanup operation is differed to the end of the request cycle and error trapped to prevent the user from getting the error page simply because this operation failed
- Session starting is deferred to a point after the application is stored to the factory; if the session fails to start for whatever reason this should put the user in an error page instead of the "Application Instantiation Error" stuff showing
- One must explicitly call `$session->start()` to correctly start the session right now; similar to other parts of the API that lazy load, the session can now do the same
- The session start event is reliant on global state (namely having a `JSession` object stored to the factory); this is resolved by passing the event dependency through the dispatcher
- Because of our metadata tracking mechanisms, right now the `loadSession()` method is also performing session garbage collection when the database handler is in use; this is a waste of cycles as the underlying PHP API has mechanisms to perform garbage collection AND handles expired sessions just fine without us beating it to the punch

These are all deep rooted architectural issues that need to be addressed and the only way to do so is testing through these technically heavy patches.

**Comment by andrepereiradasilva** [comment]
- applied patch
- in frontend deleted all cookies to start a new session
- refreshed the page.
Result: 500 error

image

Same error in the admin
image

**Comment by mbabker** [comment]
Fixed.  That last merge didn't go well at all.

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 42fe67a065ba9b7632291f97081dce6fa8c067d9<br /><br />> 1) Download the branch from https://github.com/mbabker/joomla-cms/archive/refactor-session-metadata-coupling.zip
> 2) Install Joomla; sessions in the installation application should continue to function correctly

OK

> 3) Use Joomla normally; sessions should continue to function correctly

OK

> 4) With the database session handler enabled, JApplicationCms::checkSession() should only be executed when the session is new. As this is only ensuring the extra metadata is present in the database, it is not required for every page load. All requests after the session is started will have the record in the database updated by the session handler only (or any place making an explicit query to the #__session database table). As well, the new JApplicationCms::cleanupSessionMetadata() method should never be called.

OK. Added `$this->enqueueMessage('checkSession called', 'notice');` in JApplicationCms::checkSession() to confirm.

> 5) With any other session handler enabled, JApplicationCms::checkSession() should only be called if the session is new or the time is an even numbered second. The new JApplicationCms::cleanupSessionMetadata() method should be called when the onAfterRender event is triggered and the delete query should only run when the time is an even numbered second.

Didn't test don't have other session handler to test.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10905">issues.joomla.org/joomla-cms/10905</a>.</sub>

**Comment by mbabker** [comment]
Should be able to use the filesystem (PHP, None, whatever it's called in the UI these days).

**Comment by andrepereiradasilva** [comment]
ok of course! :smile: will check that

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on 42fe67a065ba9b7632291f97081dce6fa8c067d9<br /><br />PHP, Database
No memcache<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10905">issues.joomla.org/joomla-cms/10905</a>.</sub>

**Comment by andrepereiradasilva** [comment]
Works with PHP as session handler too, with this handler the notice added appears somewhat random as supposed (`should only be called [...] or the time is an even numbered second`).

**Comment by infograf768** [comment]
Hmm, @mbabker 

Could this be the reason why I get a PHP Warning?

See https://github.com/joomla/joomla-cms/issues/12108

**Comment by infograf768** [comment]
Reverting this PR solves the issue

**Comment by brianteeman** [comment]
This should not have been merged into 3.6.3 it was not a bug fix and should have waited for 3.7

**Comment by mbabker** [comment]
Wouldn't have mattered what branch it went into.  Still would've hit the same bugs because some condition's not going to be tested through normally and it's going to result in someone getting something broken.  So, I'm not pushing it anymore.  We'll just stick with the broken global state and default structures we have now.

**Pull request merged**